### DISCUSSION TITLE
how do I implement cut off generators and loads

### DISCUSSION THREAD

**Discussion by seedpower** [CURENT/andes#393] [Q&A]
My goal is to keep transient stability:

> In practical engineering, the generators generally are cut off as a whole.

Q1: At present, I have two ways to achieve this. One is to modify the `u` status of the generator, and the other is to set `tm0=0` of the generator . Which one is better? 

> loads usually are cut off in acertain proportion.

Q2: Can I achieve this by setting `Ppf * (1-proportion)` of PQ load?

**Comment by cuihantao** [comment]
Q1: Set `u` to 0. One should not modify `tm0`.

Q2: That works. You can alternatively split the into into two when preparing the data.

**Reply by seedpower** [reply]
Thank you. The following is a generator cutting implementation.

```
import os
import andes
import copy
from andes.utils import get_case

ss = andes.run(andes.get_case("ieee14/ieee14_full.xlsx"), setup=False, default_config=True)
ss.add("Fault", dict(bus=8, tf=1.0, tc=2.0))
ss.setup()
ss.PFlow.run()

# tf=1.5s, cuf off GENROU5
ss.TDS.config.tf = 1.5
ss.TDS.config.criteria = 1
ss.TDS.run()

shed_gens = [5]
u_shed_idx = ss.GENROU.find_idx(keys='gen', values=shed_gens)
u = ss.GENROU.get(src='u', idx=u_shed_idx, attr='v')
u_new = u * 0
ss.GENROU.set(src='u', idx=u_shed_idx, attr='v', value=u_new)

# t=2.0s, restore u
ss.TDS.config.tf = 2.0
ss.TDS.run()
ss.GENROU.set(src='u', idx=u_shed_idx, attr='v', value=u)

# tf = 5s
ss.TDS.config.tf = 5
ss.TDS.config.criteria = 1
ss.TDS.run()
ss.TDS.load_plotter()
ss.TDS.plotter.plot(ss.GENROU.delta)
```

If the GENROU5 is not cut off, the transient instability occurs in t=1.99s.
![图片](https://user-images.githubusercontent.com/11022830/207239303-4d5da37d-83d7-41ac-a794-9c1889ec5901.png)

Whether it is correct, perform the generator cutting operation from t=1.5s to t=2.0s, and the GENROU.delta is not equal to 0?
![图片](https://user-images.githubusercontent.com/11022830/207239208-e1af7f0c-26f6-4510-8003-12faf79618f8.png)

**Comment by cuihantao** [comment]
The generator was successfully disconnected because the rotor angle no longer changes. The power injection equation is u*(vd*Id+vq*Iq), which is zero after turning off.

Reconnecting a generator poses a different question. The generator’s rotor angle upon reconnection will have an impact on the synchronization. Currently, the generator rotor angle remains at the pre fault value when disconnected. I don’t know how other tools handle this. If you know please post it here.

Get Outlook for iOS<https://aka.ms/o0ukef>
________________________________

**Reply by seedpower** [reply]
I don’t know. Can I use `tm0=0` to simulate it?

```
import os
import andes
import copy
from andes.utils import get_case

ss = andes.run(andes.get_case("ieee14/ieee14_full.xlsx"), setup=False, default_config=True)
ss.add("Fault", dict(bus=8, tf=1.0, tc=2.0))
ss.setup()
ss.PFlow.run()

# tf=1.5s, cuf off GENROU5
ss.TDS.config.tf = 1.5
ss.TDS.config.criteria = 0
ss.TDS.run()

shed_gens = [5]
tm0_shed_idx = ss.GENROU.find_idx(keys='gen', values=shed_gens)
tm0 = ss.GENROU.get(src='tm0', idx=tm0_shed_idx, attr='v')
tm0_new = tm0 * 0
ss.GENROU.set(src='tm0', idx=tm0_shed_idx, attr='v', value=tm0_new)

# t=2.0s, restore tm0
ss.TDS.config.tf = 2.0
ss.TDS.run()
ss.GENROU.set(src='tm0', idx=tm0_shed_idx, attr='v', value=tm0)

# tf = 5s
ss.TDS.config.tf = 5
ss.TDS.config.criteria = 0
ss.TDS.run()
ss.TDS.load_plotter()
ss.TDS.plotter.plot(ss.GENROU.delta)

```

![图片](https://user-images.githubusercontent.com/11022830/207609980-6e08b599-7db5-4f0b-b5f5-0fbddf73308a.png)

**Comment by cuihantao** [comment]
Please don't modify `tm0` in any case. Setting `tm0` to zero will not turn off the generator but provide a step up `tm` in the amount of `tm0`.

See a discussion here: https://psspy.org/psse-help-forum/question/2399/reconnecting-generators-during-dynamic-simulation/

**Reply by seedpower** [reply]
Q1: Is there a good way to implement a generator dynamic brake in ANDES?
[Adaptive_Power_System_Emergency_Control_Using_Deep_Reinforcement_Learning_2.pdf](https://github.com/cuihantao/andes/files/10229503/Adaptive_Power_System_Emergency_Control_Using_Deep_Reinforcement_Learning_2.pdf)
paper github: https://github.com/RLGC-Project/RLGC

Q2: If a generator cannot be dynamic brake in ANDES, how to implement cut off 50% of the a generator capacity dynamically?

**Reply by cuihantao** [reply]
Generator power adjustment is done through turbine governors, as in actual systems. You can alter the value of `paux0` of the turbine governor to change the power setpoints. Refer to the specific governor model to calculate a `paux0` that achieves a power reduction at your desired percentage.

**Reply by seedpower** [reply]
Thank you. I implemented a `cut_off_generator_capacity()` function to cut off desired percentage of generator capacity. 
I called the function twice to cut off 40% of generator capacity.  But I am not sure if it is right of the calculation of `paux0`. Is that right?

```
import andes

ss = andes.run(andes.get_case("ieee14/ieee14_full.xlsx"), setup=False, default_config=True)
ss.setup()
ss.PFlow.run()


def cut_off_generator_capacity(shed_gens, off_ratio):
    p_shed_idx = ss.TGOV1.find_idx(keys='syn', values=shed_gens)
    paux0 = ss.TGOV1.get(src='paux0', idx=p_shed_idx, attr='v')
    pref = ss.TGOV1.get(src='pref', idx=p_shed_idx, attr='v')

    # CHECK - calculate new paux0 according to off_ratio
    paux0_new = paux0 + pref * (-off_ratio)

    ss.TGOV1.set(src='paux0', idx=p_shed_idx, attr='v', value=paux0_new)

    print()
    print("pref\t=\t" + str(pref))
    print("paux0\t=\t" + str(paux0))
    print("paux0_new\t=\t" + str(paux0_new))
    print()

# tf = 1.0
ss.TDS.config.tf = 1.0
ss.TDS.run()
cuff_off_generator_capacity(["GENROU_1"], 0.2)
cuff_off_generator_capacity(["GENROU_1"], 0.2)

# tf = 10
ss.TDS.config.tf = 10
ss.TDS.config.criteria = 0
ss.TDS.run()
ss.TDS.load_plotter()
ss.TDS.plotter.plot(ss.GENROU.tm)

```

**Reply by cuihantao** [reply]
The calculation of `paux0` does vary between different turbine governor models.

For TGOV1 (https://www.powerworld.com/WebHelp/Content/TransientModels_PDF/Generator/Governor/Governor%20TGOV1.pdf), `paux0`, which is equal to `paux`, modifies the `pref` signal that will be multiplied by `1/R`. To get a 40% reduction in the output, your reduction has to be multiplied by `R`. 

We have a TGOV1N model which moves the `paux` to be summed at the output of `1/R`. See https://www.powerworld.com/WebHelp/Content/TransientModels_PDF/Generator/Governor/Governor%20TGOV1.pdf. 

The way you check the `tm` looks correct to me.
